qiskit/utils/optionals.py

```diff
@@ -0,0 +1,292 @@
+    msg="This may not be possible for all Python versions and OSes.",
+)
+HAS_CVXPY = LazyImportTester("cvxpy", install="pip install cvxpy")
+HAS_DOCPLEX = LazyImportTester(
+    {"docplex": (), "docplex.mp.model": ("Model",)},
+    install="pip install 'qiskit-terra[bip-mapper]'",
+    msg="This may not be possible for all Python versions and OSes.",
+)
+HAS_FIXTURES = LazyImportTester("fixtures", install="pip install fixtures")
+HAS_IPYTHON = LazyImportTester("IPython", install="pip install ipython")
+HAS_IPYWIDGETS = LazyImportTester("ipywidgets", install="pip install ipywidgets")
+HAS_JAX = LazyImportTester(
+    {"jax": ("grad", "jit"), "jax.numpy": ()},
+    name="jax",
+    install="pip install jax",
+)
+HAS_MATPLOTLIB = LazyImportTester(
+    ("matplotlib.patches", "matplotlib.pyplot"),
+    name="matplotlib",
+    install="pip install matplotlib",
+)
+HAS_NETWORKX = LazyImportTester("networkx", install="pip install networkx")
+
+
+def _nlopt_callback(available):
+    if not available:
+        return
+    import nlopt  # pylint: disable=import-error
+
+    _logger.info(
+        "NLopt version: %s.%s.%s",
+        nlopt.version_major(),
+        nlopt.version_minor(),
+        nlopt.version_bugfix(),
+    )
+
+
+HAS_NLOPT = LazyImportTester(
+    "nlopt",
+    name="NLopt Optimizer",
+    callback=_nlopt_callback,
+    msg="See the documentation of 'qiskit.algorithms.optimizer.nlopts' for installation help.",
+)
+HAS_PIL = LazyImportTester({"PIL": ("Image",)}, name="pillow", install="pip install pillow")
+HAS_PYDOT = LazyImportTester("pydot", install="pip install pydot")
+HAS_PYLATEX = LazyImportTester(
+    {
+        "pylatexenc.latex2text": ("LatexNodes2Text",),
+        "pylatexenc.latexencode": ("utf8tolatex",),
+    },
+    name="pylatexenc",
+    install="pip install pylatexenc",
+)
+HAS_SEABORN = LazyImportTester("seaborn", install="pip install seaborn")
+HAS_SKLEARN = LazyImportTester(
+    {"sklearn.linear_model": ("Ridge", "Lasso")},
+    name="scikit-learn",
+    install="pip install scikit-learn",
+)
+HAS_SKQUANT = LazyImportTester(
+    "skquant.opt",
+    name="scikit-quant",
+    install="pip install scikit-quant",
+)
+HAS_SQSNOBFIT = LazyImportTester("SQSnobFit", install="pip install SQSnobFit")
+HAS_SYMENGINE = LazyImportTester("symengine", install="pip install symengine")
+HAS_TESTTOOLS = LazyImportTester("testtools", install="pip install testtools")
+HAS_Z3 = LazyImportTester("z3", install="pip install z3-solver")
+
+HAS_GRAPHVIZ = LazySubprocessTester(
+    ("dot", "-V"),
+    name="graphviz",
+    install="'brew install graphviz' if on Mac, or by downloding it from their website",
+)
+HAS_PDFLATEX = LazySubprocessTester(
+    ("pdflatex", "-version"),
+    msg="You will likely need to install a full LaTeX distribution for your system",
+)
+HAS_PDFTOCAIRO = LazySubprocessTester(
+    ("pdftocairo", "-v"),
+    msg="This is part of the 'poppler' set of PDF utilities",
+)
+
+del LazyImportTester, LazySubprocessTester
```

Comment:
Is there a reason we delete these from the module here? It's a bit weird could we just use `__all__` instead? If it's not sufficient could you add a comment explaining why this is needed?

Reply:
It was just me being picky about not leaking names out of the module.  Anything we make public in a module _invariably_ seems to end up with people using that as the import path, so I was just tidying up the names.  I changed that to be more explicit in 897f657db by making the names private right from the import.